### DISCUSSION TITLE
Multiple calls to curl_global_init in a multi threaded environment

### DISCUSSION THREAD

**Discussion by disa6302** [curl/curl#10767] [Q&A]
Hello,

I am working on a project where I have a library that invokes `curl_global_init`. This library is being called from multiple threads which means there will be multiple invocations to `curl_global_init`. I read the documentation that curl_global_init is not thread safe until v7.84 and I am running this on a Ubuntu 22.04 which has curl 7.81 as the latest. The error I encounter while running `curl_easy_perform()` to access a URL is: `"Timeout was reached: SSL connection timeout"`. My hunch is this is happening because curl_global_init() is being invoked in each thread. The documentation is pretty vague on what kind of issues can occur because global_init is not thread safe. Has anybody run into this error and have a set up similar to mine?

What would happen if global_init is invoked multiple times?

**Comment by bagder** [comment]
The information is vague because it depends on so many different factors, 3rd party libraries and the versions of those libraries so it is not possible to be exact and cover your specific situation.

These days, with a recent OpenSSL (using thread-safe init by itself), there should be very little risk in doing multiple `curl_global_init()` calls the way you describe.

I do not think your problem is due to multiple calls to global init.

**Reply by disa6302** [reply]
@bagder ,

Thank you for getting back. The library I am talking about is our own implementation. I have a function which invokes curl_global_init(). In the higher layer application, I have 8 Java threads which use this function via the JNI. At times, the function is invoked at the same time from these threads and  often, I end up seeing the timeout error. I dont think it is network, because it is an EC2 instance and it is stable. 

Generally speaking, is it safe to invoke curl_global_init from multiple threads? I assume not based on documentation that suggests this is not thread safe. The documentation is not clear on what the implications are. 

As I mentioned earlier, I am using libcurl 7.81. Any leads on scenarios when this timeout can occur would be super helpful.
